File: app/src/main/java/com/kieran/winnipegbus/Views/RouteNumberTextView.java
package com.kieran.winnipegbus.Views;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.TextView;

import com.kieran.winnipegbus.R;
import com.kieran.winnipegbusbackend.Route;
import com.kieran.winnipegbusbackend.ScheduledStop;
import com.kieran.winnipegbusbackend.enums.CoverageTypes;

public class RouteNumberTextView extends TextView {
    public RouteNumberTextView(Context context) {
        super(context);
    }

    public RouteNumberTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public RouteNumberTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @SuppressWarnings("deprecation")
    public void setColour(ScheduledStop scheduledStop) {
        CoverageTypes coverageType = scheduledStop.getCoverageType();

        if (Route.isDownTownSpirit(scheduledStop.getRouteNumber())) {
            setTextColor(getResources().getColor(R.color.white));
            setBackgroundResource(R.drawable.route_number_background_dt_spirit);
        } else if (coverageType == CoverageTypes.REGULAR) {
            setTextColor(getResources().getColor(R.color.black));
            setBackgroundResource(R.drawable.route_number_background_regular);
        } else if (coverageType == CoverageTypes.EXPRESS || coverageType == CoverageTypes.SUPER_EXPRESS) {
            setTextColor(getResources().getColor(R.color.black));
            setBackgroundResource(R.drawable.route_number_background_express);
        } else if (coverageType == CoverageTypes.RAPID_TRANSIT) {
            setTextColor(getResources().getColor(R.color.white));
            setBackgroundResource(R.drawable.route_number_background_rt);
        }
    }
}

File: app/src/main/java/com/kieran/winnipegbus/views/RouteNumberTextView.kt
package com.kieran.winnipegbus.views

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.widget.TextView

import com.kieran.winnipegbus.R
import com.kieran.winnipegbusbackend.agency.winnipegtransit.WinnipegTransitRouteIdentifier
import com.kieran.winnipegbusbackend.agency.winnipegtransit.WinnipegTransitService
import com.kieran.winnipegbusbackend.enums.CoverageTypes
import com.kieran.winnipegbusbackend.interfaces.RouteBadge
import com.kieran.winnipegbusbackend.interfaces.RouteIdentifier
import kotlin.math.round

class RouteNumberTextView : TextView {
    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    fun setColour(routeBadge: RouteBadge?) {
        if (routeBadge == null)
            return

        setTextColor(Color.parseColor(routeBadge.getTextColour()))
        setBackgroundColor(Color.parseColor(routeBadge.getBackgroundColour()))
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        if (textSize < spToPx(40)) {
            val width = spToPx(36)
            val height = spToPx(24)
            setMeasuredDimension(width, height)
        }
    }

    private fun spToPx(dp: Int): Int {
        val density = context.resources
                .displayMetrics
                .scaledDensity
        return round(dp.toFloat() * density).toInt()
    }

}
